src/core/core/network.cpp

```diff
@@ -352,27 +369,28 @@ namespace dsn
     {
     }
 
-    bool rpc_session::on_disconnected(bool is_write)
+    bool rpc_session::on_disconnected(bool /*is_write*/)
     {
-        if (is_client())
+        if (set_disconnected())
         {
-            set_disconnected();
             rpc_session_ptr sp = this;
-            _net.on_client_session_disconnected(sp);
+            if (is_client())
+            {
+                _net.on_client_session_disconnected(sp);
+            }
+            else
+            {
+                _net.on_server_session_disconnected(sp);
+            }
+
+            clear_send_queue(false);
+
+            return true;
         }
-        
         else
         {
-            rpc_session_ptr sp = this;
-            _net.on_server_session_disconnected(sp);
-        }
-
-        if (is_write)
-        {
-            clear_send_queue(false);
```

Comment:
In my changes, the on_disconnected() will be invoked once and only once when connection failed. So, we do not distinguish "is_write" here.